File: 백준/Silver/1021. 회전하는 큐/회전하는 큐.py
import sys
from collections import deque

if __name__ == '__main__':
    N, M = map(int, sys.stdin.readline().split())

    dq = deque([i for i in range(1, N+1)])
    lst = list(map(int, sys.stdin.readline().split()))
    cnt = 0
    for n in lst:
        idx = 0
        for i in range(len(dq)):
            if dq[i] == n:
                idx = i
                break
        # idx  0 보다는 같거나 크고  len(dq)-1 보다 같거나 작다.

        ## 중간 위치 보다 크다면
        if idx > len(dq) // 2:
            # 우측 로테이션
            while dq[0] != n:
                element = dq.pop()
                dq.appendleft(element)
                cnt += 1
        else:
            # 왼쪽 로테이션
            while dq[0] != n:
                element = dq.popleft()
                dq.append(element)
                cnt += 1

        dq.popleft()

    print(cnt)

File: 백준/Silver/1021. 회전하는 큐/회전하는 큐.java
import java.io.*;
import java.util.Deque;
import java.util.LinkedList;

public class Main{
    public static void main(String[] args) throws IOException{

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        String[] inputs = br.readLine().split(" ");
        int N = Integer.parseInt(inputs[0]);
        int M = Integer.parseInt(inputs[1]);
        Deque<Integer> dq = new LinkedList<>();
        int[] positions = new int[M];
        for (int i=1; i<=N; i++) {
            dq.add(i);
        }

        inputs = br.readLine().split(" ");
        for (int i=0; i<M; i++) {
            positions[i] = Integer.parseInt(inputs[i]);
        }

        int cnt = 0;
        for (int position : positions) {
            int idx = 0;
            for(int i : dq) {
                if (position == i) {
                    break;
                }
                idx++;
            }

            if (idx > dq.size() / 2) {
                while (dq.peek() != position) {
                    dq.addFirst(dq.pollLast());
                    cnt++;
                }
            } else {
                while (dq.peek() != position) {
                    dq.addLast(dq.pollFirst());
                    cnt++;
                }
            }
            dq.poll();
        }

        bw.write(cnt+"\n");
        bw.flush();
    }
}